heliopy/data/mms.py

```diff
@@ -7,131 +7,1781 @@
+            data_root = pathlib.Path(heliopy.config['download_dir']) / 'mms'
+
+        self.data_root = data_root
+        self._sdc_home = 'https://lasp.colorado.edu/mms/sdc'
+        self._info_type = 'download'
+
+        # Create a persistent session
+        self._session = requests.Session()
+        self._session.auth = (sdc_username, sdc_password)
+
+    def __str__(self):
+        return self.url()
+
+    # https://stackoverflow.com/questions/17576009/python-class-property-use-setter-but-evade-getter
+    def __setattr__(self, name, value):
+        """Control attribute values as they are set."""
+
+        # TYPE OF INFO
+        #   - Unset other complementary options
+        #   - Ensure that at least one of (download | file_names |
+        #     version_info | file_info) are true
+        if name == 'anc_product':
+            self.data_type = 'ancillary'
+
+        elif name == 'data_type':
+            if 'gls_selections' in value:
+                if value[15:] not in ('mp-dl-unh',):
+                    raise ValueError('Unknown GLS Selections type.')
+            elif value not in ('ancillary', 'hk', 'science',
+                               'abs_selections', 'sitl_selections'
+                               ):
+                raise ValueError('Invalid value {} for attribute'
+                                 ' "{}".'.format(value, name))
+
+            # Unset attributes related to data_type = 'science'
+            if 'selections' in value:
+                self.sc = None
+                self.instr = None
+                self.mode = None
+                self.level = None
+                self.optdesc = None
+                self.version = None
+
+        elif name == 'files':
+            if value is not None:
+                self.sc = None
+                self.instr = None
+                self.mode = None
+                self.level = None
+                self.optdesc = None
+                self.version = None
+
+        elif name == 'level':
+            # L2 and L3 are the only public data levels
+            if value in [None, 'l2', 'l3']:
+                self.site = 'public'
+            else:
+                self.site = 'private'
+
+        elif name == 'site':
+            # Team site is most commonly referred to as the "team",
+            # or "private" site, but in the URL is referred to as the
+            # "sitl" site. Accept any of these values.
+            if value in ['private', 'team', 'sitl']:
+                value = 'sitl'
+            elif value == 'public':
+                value = 'public'
+            else:
+                raise ValueError('Invalid value for attribute {}.'
+                                 .format(name)
+                                 )
+
+        elif name in ('start_date', 'end_date'):
+            # Convert string to datetime object
+            if isinstance(value, str):
+                try:
+                    value = dt.datetime.strptime(value[0:19],
```

Comment:
SunPy has a helpful function `parse_time` that will allow a time to be given in lots of different formats, so it might be worth using that here and below. http://docs.sunpy.org/en/stable/api/sunpy.time.parse_time.html?highlight=parse_time#sunpy.time.parse_time